### ISSUE TITLE
Pure JS implementation of Michelson parsing / formatting / validation

### ISSUE THREAD

**Issue by simonmcl** [ecadlabs/taquito#441]
I'm writing a native app in Swift for Tezos and would love the ability to have Michelson parsing / validating support in our app. We can't run typescript in Swift or Kotlin, but there is limited support for JS in both platforms, can't run ajax calls, but can execute standalone functions.

Would it be possible for you guys to setup a CI/CD process to generate a JS file from the Michelson codec and publish it as a github release, as a standalone file. I'd love to try take that and incorporate it into a native Swift / Kotlin library, and update it as you add more features / bug fixes etc.

**Comment by jevonearth** [comment]
Hi @simonmcl 

You should be able to use the NPM package `@taquito/michel-codec` as is. Available here: https://www.npmjs.com/package/@taquito/michel-codec

This package is updated, along with all the Taquito packages on every release. Let us know if you run into any issues!

**Comment by simonmcl** [comment]
@jevonearth sorry I might not have been clear. So Natively (on iOS using `JSContext`) you can only use a single static JS file. You can't pass in an NPM module with .ts files. I've seen talk online of users having to run tools like browserify and/or webpack to convert NPM modules into static JS files.

There can be no:

```
import { Prim, Expr, IntLiteral } from “./micheline”;
import { Tuple, NoArgs, ReqArgs, NoAnnots } from “./utils”;
```

Inside the files in order for it to work out of the box. Think like a minified js file coming from a CDN

**Comment by jevonearth** [comment]
Hi @simonmcl 

Can you tell me a bit more about your environment? I infer that you are writing an iOS app in Swift? and you are wanting to use `michel-codec` within that context?

It might be a bit much to ask, but if you can provide a tiny example project, that would probably help us a lot to nail this use-case. If that's too much effort, no worries, we will figure something out.

**Comment by simonmcl** [comment]
@jevonearth Hi, I looked at this again as I badly needed the local-forging package, I have a solution using webpack thats working for that package, it would be great if you could integrate it into all the packages and set this up as part of a CI tool.

**The problem:**
Native support for javascript is limited, and must be inside a single file. While the current build commands generate `xxx.es5.js` and `xxx.umd.js` which are supposed to be universal, they still include imports and requires, linking to other packages. Some are internal to Node.js and some are other taquito packages. These files are unusable to mobile devs as inside a native app you don't have the node libraries, and a js file can't break outside the context created for it, nor can it access the disk.

**The solution:** 
After a lot of banging my head against the wall re-learning webpack, I was able to export the taquito-local-forging package into a single js file, compatible with iOS using the following script, which I added to a local copy of the code, named `webpack.config.js`:

```Javascript
const webpack = require('webpack');
const path = require('path');

module.exports = {
    entry: {
        "taquito_local_forging": ['./src/taquito-local-forging.ts']
    },
    mode: 'production',
    module: {
        rules: [
            {
                test: /\.ts$/,
                use: 'ts-loader',
                exclude: /node_modules/
            }
        ]
    },
    devtool: 'source-map',
    resolve: {
        extensions: ['.tsx', '.ts', '.js'],
        modules: ['node_modules'],
        fallback: { 
            "stream": require.resolve("stream-browserify") 
        }
    },
    output: {
        filename: '[name].js',
        path: path.resolve(__dirname, 'dist'),
        library: ['[name]'],
        libraryTarget: "var"
    },
    plugins: [
        new webpack.ProvidePlugin({ Buffer: ['buffer', 'Buffer'] })
    ]
};
```

Then modified package.json scripts to include: 

```
"build-webpack": "webpack --progress --color",
```

and ran:

```
npm run build-webpack
```

It would be great if you could add a file like this to all the packages, and maybe one for the root as well, an run them through Travis CI or something similar for every release. A way to test it would be adding it to a HTML file inside a script tag and seeing if the basic functionality can be accessed.

It took a long time to download all the tools and get local builds working. It would be much easier for mobile devs if they could just download the release zip and take the .js file out that they need (maybe name them differently from the `es5` / `umd` files).

Dealing with the Promise objects in swift is also quite tricky. I'll try write something about how to use it when I get this cleaned up on my side, as this is all very badly documented across the platforms.

**Comment by jevonearth** [comment]
Thanks @simonmcl, sorry for the silence on this issue. We will get it into our planning cycle.

**Comment by roxaneletourneau** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Taquito-Development-Team-60ca841c87ca580010057707/issues/ecadlabs/taquito/441?planning-poker) with ZenHub @dsawali @hui-an-yang @zainen

**Comment by dsawali** [comment]
Hello @simonmcl, I'm currently taking a look at this issue. Thank you again for providing a POC for us to work with, I think we can adapt that for some of our packages in Taquito. 

Could you state which packages you would like to be bundled as js files again? While it might be ideal for you to do all packages, we're thinking of rolling out a few packages first as a trial.

**Comment by simonmcl** [comment]
@dsawali i'm currently only using the local-forging package. I can't use anything that has networking requests or anything like that in it

**Comment by dsawali** [comment]
@simonmcl Sounds good. Just to confirm what you really need, would such a js file be available in the `dist` folder be enough to be helpful to your development process?

Also since I'm not an iOS or Android dev, do you have any suggestions on how we could test whether the bundled js file works as intended for our users?

**Comment by simonmcl** [comment]
@dsawali I think the ideal solution would be to have a zip of all the native-mobile-compatible js files appear as a "asset" on each github release (see pic). Setting up an alert for new releases and then just downloading the file is going to be easiest for devs.

![Screenshot 2022-07-15 at 09 35 45](https://user-images.githubusercontent.com/8058386/179186358-57c4d9e6-8e18-494d-94d0-7c01496e0e75.png)
 
Native mobile works much the same as a standard browser without any advanced/modern tooling, but its headless. There is no root `this`, `window` etc. A JS library has to be wrapped up into a variable that it defines. The config I gave does this and creates a top level variable called `taquito_local_forging`, which contains everything inside it.

If it can be imported into raw HTML and used with basic javascript script tags, outside of Node.js, something like the below, it should be fine for mobile

```
<script type="text/javascript" src="taquito-local-forging.js"></script>
<script type="text/javascript">
     var forger = new taquito_local_forging.LocalForger();
     forger.forge(...)
</script>
```

**Comment by dsawali** [comment]
@simonmcl Thank you for the feedback. I'll do some testing and see what we can do to accommodate that.

**Comment by dsawali** [comment]
Hello @simonmcl, sorry for the delay. We currently have a solution we're brewing. But before we implement a new webpack bundler as a dependency, have you tried using the Taquito CDN?

https://www.npmjs.com/package/@taquito/taquito

A quick rundown of it is basically to import the CDN as you would a script as such:
```
  <script src="https://unpkg.com/@taquito/taquito@13.0.1/dist/taquito.min.js" crossorigin="anonymous" integrity="sha384-78JSzLWa1w1NqUK5WcwaalS71hUZFjtTsSzh2af8arFgioOMhzZvvsRB91NvHmxa">  </script>
```

in the same file, you could access the Taquito functions as such:
```
var taquito = window.taquito;
```
which will give you access to Taquito functions such as the LocalForger. If you're wondering what functions/classes are available in it, you can simply `console.log(taquito)`

Could you confirm whether the CDN is a good alternative to the fork you made using webpack?

**Comment by simonmcl** [comment]
@dsawali no that won't work. Looks like you've not read my previous comment.

Native mobile JS doesn't have access to "window", as it's a headless JS implementation. It all has to be bundled up into a single Js file, without any imports, exports, or links to other files in anyway. All self contained, and wrapped up inside a root variable declared in the JS. 

Please have a look at the JS file that was created by the web pack config as an example

**Comment by dsawali** [comment]
@simonmcl I must have missed that, pardon me. Thanks for confirming. 

We've taken a look at your Webpack config example, and we are planning to put that into our workflow to accommodate mobile developers in the community. I just thought that in the meantime, the CDN might be helpful before the change comes about in the next release. In the upcoming release, you should have taquito-local-forging available as a downloadable zip file in each release tag. We will also update the documentation to reflect the change we make.

**Comment by roxaneletourneau** [comment]
A JS bundle of the taquito-local-forging for the latest release (14.0.0) is available here: https://github.com/ecadlabs/taquito/releases/tag/14.0.0

Keeping the issue open to evaluate the need to generate such bundle for the other packages.

**Comment by simonmcl** [comment]
@roxaneletourneau apologies for only looking at this now. I've tested the latest version and yes it works great just copy / pasting it into my Swift project

It would be great if this could be expanded to any other areas that don't require network or complex imports. For example, the code around `valueDecoder` and `emitMicheline` is used to make human readable versions of the sign payload coming from Beacon. This is another thing that I need. But currently this is spread out over 2 packages, and the `valueDecoder` from the local-forging package doesn't show up in the vanilla js (probably because the source is in another folder I might have missed with my script). If this is something you guys could add in some way, i'd hugely appreciate it

**Comment by dsawali** [comment]
Hello @simonmcl, have you tried using the latest release of the taquito-local-forging bundle? https://github.com/ecadlabs/taquito/releases/tag/17.0.0

It should be uploaded under 'assets'.
> valueDecoder from the local-forging package doesn't show up in the vanilla js (probably because the source is in another folder I might have missed with my script)
that's interesting, the `valueDecoder` method is created in the same source, but maybe it isn't exported in the file that you need. I'll try and re-export that method and recreate the zip file for you to test some time in the near future so you can test it out

**Comment by simonmcl** [comment]
@dsawali yes It was v17 that I used last time. I now have that working in my native iOS code. You can see if you download and unzip it, and open it in a text editor. Do search for "LocalForger" and you will see thats present, do a search for "valueDecoder" and its not. Maybe the code might be present in some form, but the export names aren't

Getting valueDecoder is only half the problem though. Ideally I need valueDecoder from that package and `emitMicheline` from the other. Then i'll need to find a way to combine them on my side

**Comment by dsawali** [comment]
yeah that's fine, I can try and just export those methods in local forger and see if the preview package works for you. Hang tight @simonmcl

**Comment by oldominion** [comment]
I have a similar use case to this. I need functionality of taquito and beacon-wallet for use in <script> tags, but currently only taquito project produces a .min.js file currently, which doesn't include beacon-wallet functions.
